packages/web3-providers-ws/test/integration/ganache_fault_tolerance.test.ts

```diff
@@ -0,0 +1,389 @@
+			const connectPromise = new Promise(resolve => {
+				webSocketProvider.once('connect', () => {
+					mockCallback();
+					resolve(true);
+				});
+			});
+			await connectPromise;
+			//
+			webSocketProvider.disconnect();
+			await waitForEvent(webSocketProvider, 'disconnect');
+
+			webSocketProvider.connect();
+			const connectPromise2 = new Promise(resolve => {
+				webSocketProvider.once('connect', () => {
+					mockCallback();
+					resolve(true);
+				});
+			});
+			await connectPromise2;
+			webSocketProvider.disconnect();
+			expect(mockCallback).toHaveBeenCalledTimes(2);
+			await server.close();
+		});
+
+		it('webSocketProvider supports subscriptions', async () => {
+			const server = ganache.server();
+			await server.listen(port);
+			const webSocketProvider = new WebSocketProvider(host);
+
+			await waitForOpenSocketConnection(webSocketProvider);
+			expect(webSocketProvider.supportsSubscriptions()).toBe(true);
+
+			webSocketProvider.disconnect();
+			await server.close();
+		});
+
+		it('times out when server is closed', async () => {
+			const server = ganache.server();
+			await server.listen(port);
+			const reconnectionOptions = {
+				delay: 100,
+				autoReconnect: false,
+				maxAttempts: 1,
+			};
+			const webSocketProvider = new WebSocketProvider(host, {}, reconnectionOptions);
+			const mockCallBack = jest.fn();
+			const errorPromise = new Promise(resolve => {
+				webSocketProvider.on('error', err => {
+					if (err?.message.startsWith('connect ECONNREFUSED')) {
+						mockCallBack();
+						resolve(true);
+					}
+				});
+			});
+			await server.close();
+			await errorPromise;
+			expect(mockCallBack).toHaveBeenCalled();
+		});
+
+		it('with reconnect on, will try to connect until server is open then close', async () => {
+			const reconnectionOptions = {
+				delay: 10,
+				autoReconnect: true,
+				maxAttempts: 100,
+			};
+			const webSocketProvider = new WebSocketProvider(host, {}, reconnectionOptions);
+
+			const mockCallback = jest.fn();
+			const connectPromise = new Promise(resolve => {
+				webSocketProvider.on('connect', () => {
+					mockCallback();
+					resolve(true);
+				});
+			});
+
+			const server = ganache.server();
+			await server.listen(port);
+			await connectPromise;
+			webSocketProvider.disconnect();
+			await server.close();
+			expect(mockCallback).toHaveBeenCalledTimes(1);
+		});
+
+		it('allows disconnection on lost connection, when reconnect is enabled', async () => {
+			const reconnectionOptions = {
+				delay: 10,
+				autoReconnect: true,
+				maxAttempts: 100,
+			};
+			const webSocketProvider = new WebSocketProvider(host, {}, reconnectionOptions);
+
+			const connectPromise = waitForOpenSocketConnection(webSocketProvider);
+
+			const server = ganache.server();
+			await server.listen(port);
+			await connectPromise;
+			webSocketProvider.disconnect();
+			expect(!!(await waitForEvent(webSocketProvider, 'disconnect'))).toBe(true);
+			await server.close();
```

Comment:
Shouldn't `await server.close()` happen before `webSocketProvider.disconnect();` since we're testing `allows disconnection on lost connection`?

Reply:
Good catch, thanks! Changed